List buyed albums by user

gabrielbori opened the pull request:
## Summary

GET "/users/:user_id/albums"
User must be logged. If it is admin, it can get any user albums. If user not admin, it can only get buyed albums by itself.

## Known Issues


## Trello Card

gabrielbori committed:
Users list (#12)

* Added users middleware, post for users signUp

* add endpoint and validations to create users

* deleted empty file

* Add log when field valitation fails

* fixed some checks in user middleware, add error handling

* Fix some names

* Fix error, will change un other branch

* tests

* Update user.js

* Add migrations tests

* Finished signUp tests

* Added test migrations, deleted extra file

* Working on signIn feature, added token response

* Add enter, delete extra primise.resolve

* Update signUp.js

* Update users.js

* Fix serveral tests. Fix error messages

* Deleted coments

* Added SECRET to travis file

* Update .travis.yml

* Added middleware logic. Added token verification. Still need some changes

* Added variable validUser to reuse in tests

* Changed test to use validuser fields

* Fix errors in user controller

* Changed error handling

* changes in validations and errors

* Delete console.log

* changes in constants and error messages. Added helpers folder

* Add created user serializer

* Now using funcion to creaate user

* Changed logic to test email existance, add mapper to hash password

* Changed logic to test email existance, add mapper to hash password

* fix error in mapper

* fix error in mapper

* Changed constants to use snake_case

* re-doing tests with new changes from signUp

* Changed messages from param validator

* changed email already exists message

* Update users.js

* modifing tests

* Modify tests to match new error formatt

* Fixed last test to match new error handling

* change logic to work with schemas and helpers

* fix tests

* todo

* Moved eslint disable, deleted paramErrors empty object2

* Added try catch to db calls

* Fix error in returning userData when created

* Changed constants to env file. created serializer for param errors

* Changed constants to env file. created serializer for param errors

* solved require issue

* solved problem with serialize function call

* Changed env variables to use config idex and testing

* Fixes in email check error

* changed email already exists test

* changed name of findBy email function

* Pretier functions. Added serializers and enhance errors

* Fixed tests

* Changed email registered error handle

* Chaanged test to match new email error handler

* Changed error function when email does not exist. Added log

* added first test to get users list

* deleted commented function

* Added factory-girl use

* Added more tests

* Added some tests and fixed other merge issues

* Finished endpoint to get users list being authorized

* change log call

* Changed user service to services folder because could not run npm console

* Users list pr modifications

* Users list pr modifications

* Renamed variables

gabrielbori committed:
Users list (#12)

* Added users middleware, post for users signUp

* add endpoint and validations to create users

* deleted empty file

* Add log when field valitation fails

* fixed some checks in user middleware, add error handling

* Fix some names

* Fix error, will change un other branch

* tests

* Update user.js

* Add migrations tests

* Finished signUp tests

* Added test migrations, deleted extra file

* Working on signIn feature, added token response

* Add enter, delete extra primise.resolve

* Update signUp.js

* Update users.js

* Fix serveral tests. Fix error messages

* Deleted coments

* Added SECRET to travis file

* Update .travis.yml

* Added middleware logic. Added token verification. Still need some changes

* Added variable validUser to reuse in tests

* Changed test to use validuser fields

* Fix errors in user controller

* Changed error handling

* changes in validations and errors

* Delete console.log

* changes in constants and error messages. Added helpers folder

* Add created user serializer

* Now using funcion to creaate user

* Changed logic to test email existance, add mapper to hash password

* Changed logic to test email existance, add mapper to hash password

* fix error in mapper

* fix error in mapper

* Changed constants to use snake_case

* re-doing tests with new changes from signUp

* Changed messages from param validator

* changed email already exists message

* Update users.js

* modifing tests

* Modify tests to match new error formatt

* Fixed last test to match new error handling

* change logic to work with schemas and helpers

* fix tests

* todo

* Moved eslint disable, deleted paramErrors empty object2

* Added try catch to db calls

* Fix error in returning userData when created

* Changed constants to env file. created serializer for param errors

* Changed constants to env file. created serializer for param errors

* solved require issue

* solved problem with serialize function call

* Changed env variables to use config idex and testing

* Fixes in email check error

* changed email already exists test

* changed name of findBy email function

* Pretier functions. Added serializers and enhance errors

* Fixed tests

* Changed email registered error handle

* Chaanged test to match new email error handler

* Changed error function when email does not exist. Added log

* added first test to get users list

* deleted commented function

* Added factory-girl use

* Added more tests

* Added some tests and fixed other merge issues

* Finished endpoint to get users list being authorized

* change log call

* Changed user service to services folder because could not run npm console

* Users list pr modifications

* Users list pr modifications

* Renamed variables

gabrielbori committed:
Finished tests

gabrielbori committed:
Admin user support.

gabrielbori committed:
Finished tests

gabrielbori committed:
Admin user support.

gabrielbori committed:
Finished tests

gabrielbori committed:
Finished tests

gabrielbori committed:
Admin user support.

gabrielbori committed:
Admin user support.

gabrielbori committed:
Tests

gabrielbori committed:
fix merge problems

gabrielbori committed:
fix merge problems

gabrielbori committed:
fix merge and rebase problems

gabrielbori committed:
Update errorsMessages.js

gabrielbori committed:
Update errorsMessages.js

gabrielbori committed:
Fixed again some merge and rebase issue, hope it's last time·

gabrielbori committed:
Fixed again some merge and rebase issue, hope it's last time·

gabrielbori committed:
Finished buy album logic

gabrielbori committed:
Finished buy album logic

gabrielbori committed:
Changed auth way to authenticate

gabrielbori committed:
Changed auth way to authenticate

gabrielbori committed:
Added roles to admin user, modified tests

gabrielbori committed:
Added roles to admin user, modified tests

gabrielbori committed:
pr fixes

gabrielbori committed:
pr fixes

gabrielbori committed:
deleted comment

gabrielbori committed:
deleted comment

gabrielbori committed:
Delete password from user attatched to request

gabrielbori committed:
Delete password from user attatched to request

gabrielbori committed:
Started with tests, minor fixes, rebased

gabrielbori committed:
Almost finished tests, some errors ocurring while runing them

gabrielbori committed:
fixed Tests

gabrielbori committed:
merged with master

gabrielbori committed:
Changed process env url search

gabrielbori committed:
finished some fixes

gabrielbori committed:
added done

gabrielbori committed:
Fixed pr issues

gabrielbori committed:
Changed order to buy album. Refactor album by user model, took out id

gabrielbori committed:
Created logic to get albums buyed by user id

gabrielbori committed:
Try other thing

gabrielbori committed:
Added dones to see if travis stops throwing errors

gabrielbori committed:
Try other thing

gabrielbori committed:
took out done

gabrielbori committed:
finished tests

gabrielbori committed:
added catch to see error on tests

gabrielbori committed:
Changed emal createion to avoid chance error

martinremis commented on app/middlewares/auth.js:
```diff
@@ -6,7 +6,7 @@ const config = require('../../config');
 const { roles } = require('../constants/roles');
 const mappers = require('../mappers/user');
-
+const constants = require('../constants');
```
The empty line

martinremis commented on app/controllers/albums.js:
```diff
@@ -1,9 +1,9 @@
+const db = require('../models');
 const albumInteractor = require('../interactors/albums');
-
+const { serializeAlbumsResponse } = require('../serializers/albums');
```
Don't delete the empty line after constants.

martinremis commented on app/serializers/albums.js:
```diff
@@ -0,0 +1,9 @@
+exports.serializeAlbumsResponse = findedAlbums => {
```
Always try not override the object on a serializer or mapper. Just return a new object, you don't even need a constant to do it (you will need a constant in the controller)

martinremis reviewed (changes requested):
Not fully evaluated, I may add new comments :stuck_out_tongue:

gabrielbori committed:
Added empty lines, changed serializer response of albums buyed by user

gabrielbori committed:
Buy albums (#15)

* Admin user support.

* Finished tests

* Admin user support.

* Finished tests

* fix merge problems

* Update errorsMessages.js

* Fixed again some merge and rebase issue, hope it's last time·

* Changed auth way to authenticate

* Added roles to admin user, modified tests

* pr fixes

* deleted comment

* Delete password from user attatched to request

* Users list (#12)

* Added users middleware, post for users signUp

* add endpoint and validations to create users

* deleted empty file

* Add log when field valitation fails

* fixed some checks in user middleware, add error handling

* Fix some names

* Fix error, will change un other branch

* tests

* Update user.js

* Add migrations tests

* Finished signUp tests

* Added test migrations, deleted extra file

* Working on signIn feature, added token response

* Add enter, delete extra primise.resolve

* Update signUp.js

* Update users.js

* Fix serveral tests. Fix error messages

* Deleted coments

* Added SECRET to travis file

* Update .travis.yml

* Added middleware logic. Added token verification. Still need some changes

* Added variable validUser to reuse in tests

* Changed test to use validuser fields

* Fix errors in user controller

* Changed error handling

* changes in validations and errors

* Delete console.log

* changes in constants and error messages. Added helpers folder

* Add created user serializer

* Now using funcion to creaate user

* Changed logic to test email existance, add mapper to hash password

* Changed logic to test email existance, add mapper to hash password

* fix error in mapper

* fix error in mapper

* Changed constants to use snake_case

* re-doing tests with new changes from signUp

* Changed messages from param validator

* changed email already exists message

* Update users.js

* modifing tests

* Modify tests to match new error formatt

* Fixed last test to match new error handling

* change logic to work with schemas and helpers

* fix tests

* todo

* Moved eslint disable, deleted paramErrors empty object2

* Added try catch to db calls

* Fix error in returning userData when created

* Changed constants to env file. created serializer for param errors

* Changed constants to env file. created serializer for param errors

* solved require issue

* solved problem with serialize function call

* Changed env variables to use config idex and testing

* Fixes in email check error

* changed email already exists test

* changed name of findBy email function

* Pretier functions. Added serializers and enhance errors

* Fixed tests

* Changed email registered error handle

* Chaanged test to match new email error handler

* Changed error function when email does not exist. Added log

* added first test to get users list

* deleted commented function

* Added factory-girl use

* Added more tests

* Added some tests and fixed other merge issues

* Finished endpoint to get users list being authorized

* change log call

* Changed user service to services folder because could not run npm console

* Users list pr modifications

* Users list pr modifications

* Renamed variables

* Finished buy album logic

* Admin user support.

* Finished tests

* Admin user support.

* Finished tests

* Users list (#12)

* Added users middleware, post for users signUp

* add endpoint and validations to create users

* deleted empty file

* Add log when field valitation fails

* fixed some checks in user middleware, add error handling

* Fix some names

* Fix error, will change un other branch

* tests

* Update user.js

* Add migrations tests

* Finished signUp tests

* Added test migrations, deleted extra file

* Working on signIn feature, added token response

* Add enter, delete extra primise.resolve

* Update signUp.js

* Update users.js

* Fix serveral tests. Fix error messages

* Deleted coments

* Added SECRET to travis file

* Update .travis.yml

* Added middleware logic. Added token verification. Still need some changes

* Added variable validUser to reuse in tests

* Changed test to use validuser fields

* Fix errors in user controller

* Changed error handling

* changes in validations and errors

* Delete console.log

* changes in constants and error messages. Added helpers folder

* Add created user serializer

* Now using funcion to creaate user

* Changed logic to test email existance, add mapper to hash password

* Changed logic to test email existance, add mapper to hash password

* fix error in mapper

* fix error in mapper

* Changed constants to use snake_case

* re-doing tests with new changes from signUp

* Changed messages from param validator

* changed email already exists message

* Update users.js

* modifing tests

* Modify tests to match new error formatt

* Fixed last test to match new error handling

* change logic to work with schemas and helpers

* fix tests

* todo

* Moved eslint disable, deleted paramErrors empty object2

* Added try catch to db calls

* Fix error in returning userData when created

* Changed constants to env file. created serializer for param errors

* Changed constants to env file. created serializer for param errors

* solved require issue

* solved problem with serialize function call

* Changed env variables to use config idex and testing

* Fixes in email check error

* changed email already exists test

* changed name of findBy email function

* Pretier functions. Added serializers and enhance errors

* Fixed tests

* Changed email registered error handle

* Chaanged test to match new email error handler

* Changed error function when email does not exist. Added log

* added first test to get users list

* deleted commented function

* Added factory-girl use

* Added more tests

* Added some tests and fixed other merge issues

* Finished endpoint to get users list being authorized

* change log call

* Changed user service to services folder because could not run npm console

* Users list pr modifications

* Users list pr modifications

* Renamed variables

* fix merge problems

* Update errorsMessages.js

* Fixed again some merge and rebase issue, hope it's last time·

* Finished buy album logic

* Changed auth way to authenticate

* Added roles to admin user, modified tests

* pr fixes

* deleted comment

* Delete password from user attatched to request

* Almost finished tests, some errors ocurring while runing them

* fixed Tests

* Changed process env url search

* finished some fixes

* added done

* Fixed pr issues

* Changed order to buy album. Refactor album by user model, took out id

Added empty lines, changed serializer response of albums buyed by user

Pulled from master

gabrielbori committed:
Merge branch 'list-buyed-albums-by-userId' of https://github.com/gaboxxz/gb-express-js into list-buyed-albums-by-userId

gabrielbori committed:
merge

emanuelcasco commented on test/getAlbumsBuyedByUser.js:
```diff
@@ -0,0 +1,183 @@
+const app = require('../server');
+const supertest = require('supertest');
+const { factory } = require('factory-girl');
+const errors = require('../app/errors');
```
As a recomendation, when you order your required statements as follows:

1) Native modules, ie: 'fs', 'path', etc.
2) Third party modules, ie: 'supertest', 'factory-girl', etc.
3) Local modules, ie: '../app/errors', '../server', etc.

For example:
```javascript
const fs = require('fs');
const path = require('path');

const supertest = require('supertest');
const { factory } = require('factory-girl');

const app = require('../server');
```

emanuelcasco commented on test/getAlbumsBuyedByUser.js:
```diff
@@ -0,0 +1,183 @@
+  let validUserToken = null;
+  let validAdminUserToken = null;
+  beforeEach(done =>
+    factory
```
there are too much things going on here, create helpers to abstract this logic

emanuelcasco commented on test/getAlbumsBuyedByUser.js:
```diff
@@ -0,0 +1,183 @@
+      )
+      .catch(err => {
+        // eslint-disable-next-line no-console
+        console.log(err);
```
Why are you doing this? If you fail initialising test scenario you should mark tests as failed:
```javascript
.catch(err => {
  return done(err);
});
```

emanuelcasco commented on test/getAlbumsBuyedByUser.js:
```diff
@@ -0,0 +1,183 @@
+        expect(res.body.message).toBe(constants.notAdminToGetAlbums);
+        done();
+      }));
+});
```
All beforeEach steps are necessary for all this tests? If not, divide tests in different `describes` and group tests around them

emanuelcasco commented on test/getUsers.js:
```diff
@@ -30,6 +31,8 @@ describe('Get /users', () => {
           )
       )
+      // eslint-disable-next-line no-console
+      .catch(console.log)
```
Same here, if hook fails you should reject tests.

emanuelcasco reviewed (changes requested):
Avoid console.logs in your code, even for your tests

gabrielbori committed:
Modified tests with required changes in pr
